src/Hangfire.Tags.PostgreSql/PostgreSqlTransactionRetryPolicy.cs

```diff
@@ -0,0 +1,58 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Hangfire.Logging;
+using Hangfire.Tags.Storage;
+using Npgsql;
+
+namespace Hangfire.Tags.PostgreSql
+{
+    /// <summary>
+    /// Retry policy to address https://github.com/face-it/Hangfire.Tags/issues/35
+    /// </summary>
+    public class PostgreSqlTransactionRetryPolicy : ITransactionRetryPolicy
+    {
+        private readonly int? _timeBetweenRetries;
+        private readonly int _retryCount;
+        private readonly ILog _logger = LogProvider.For<PostgreSqlTransactionRetryPolicy>(); 
+
+        public PostgreSqlTransactionRetryPolicy(int retryCount, int? timeBetweenRetries = null)
+        {
+            if (retryCount < 0)
+                throw new ArgumentException("Retrycount should be 0 or higher", nameof(retryCount));
+            if (timeBetweenRetries < 0)
+                throw new ArgumentException("Time between retries should be null, 0 or higher", nameof(timeBetweenRetries));
+
+            _timeBetweenRetries = timeBetweenRetries;
+            _retryCount = 1 + retryCount;
+        }
+        
+        private static readonly HashSet<string> ConcurrencyErrorCodes = new HashSet<string>
+        {
+            "40001", //serialization_failure
+            "40P01"  //deadlock_detected 
+        };
+
+        public void RetryOnTransactionError(Action action)
+        {
+            for (var i = 0; i < _retryCount; i++)
+            {
+                try
+                {
+                    action();
+                }
+                catch (PostgresException e) when (ConcurrencyErrorCodes.Contains(e.SqlState))
+                {
+                    if (i == _retryCount - 1)
+                        throw;
+
+                    if (_logger.IsTraceEnabled())
+                        _logger.TraceException($"Operation failed with concurrency error. Retrying {i + 1}/{_retryCount + 1}", e);
+
+                    if (_timeBetweenRetries.HasValue)
+                        Thread.Sleep(_timeBetweenRetries.Value);
```

Comment:
@erwin-faceit 
I agree that `Thread.Sleep` might be a nice improvement. In addition, I suggest selecting waiting interval randomly to decrease contention probability.

Imagine a scenario when multiple threads fall victim to concurrency errors at the same time, then all of them wait for the same amount of time, and then simultaneously repeat the operation causing concurrency issues for themselves again.

Thoughts? If you see that being useful, I can implement that improvement later today.

Reply:
I was actually just thinking about that as well. The TimeBetweenRetries should probably be used as the maximum value for a randomizer.

Also, the integration test fails for me consequently, but I think that might be processor dependent. My machine is running using 16 cores for this, which will span 15 workers, creating loads of jobs each at nearly the same time. 

A solution could be to let the amount of processor have a maximum, limit the amount of jobs per worker, or make  the timeout processor dependent. The last option sounds to me like the test is going to take a lot longer on my machine though.

Reply:
> Also, the integration test fails for me consequently
When we do implement randomized waits, it should make the logic and tests more reliable.

Also, I like the idea of limiting the amount of jobs per worker or globally, and probably discarding the test timeout completely.

Another option is measuring  `successful`/`failed` job creation ratio. I'm not sure about the formulae yet, but it should probably depend on the CPU count. And so, the test would be considered failed if that ratio is below `90%` or so. Since contention issues depend on a lot of different factors, it is hard to guarantee that your jobs are enqueued with 100% probability (unless the retry policy has _infinite_ number of retries). So it makes sense to bring a small degree of fault-tolerance to the unit-test.